R/oauth-cache.R

```diff
@@ -0,0 +1,245 @@
+## nothing there is exported, so copied over, then it evolved
+
+# cache --------------------------------------------------------------
+
+gargle_default_oauth_cache_path <- "~/.R/gargle/gargle-oauth"
+
+## this is the cache setup interface for the Gargle2.0 class
+cache_establish <- function(cache = getOption("gargle.oauth_cache")) {
+  if (length(cache) != 1) {
+    stop("Cache should be length 1 vector", call. = FALSE)
+  }
+  if (!is.logical(cache) && !is.character(cache)) {
+    stop("Cache must either be logical or string (file path)")
+  }
+
+  # If NA, propose default cache file
+  # request permission to create it, if doesn't exist yet
+  # store result of that ask (TRUE or FALSE) in the option
+  if (is.na(cache)) {
+    cache <- cache_available(gargle_default_oauth_cache_path)
+    options("gargle.oauth_cache" = cache)
+  }
+  ## cache is now TRUE, FALSE or path
+
+  if (isFALSE(cache)) return()
+
+  if (isTRUE(cache)) {
+    cache <- gargle_default_oauth_cache_path
+  }
+
+  if (!file.exists(cache)) {
+    cache_create(cache)
+  }
+  ## cache is now NULL or path to a file that exists (possibly empty)
+
+  return(cache)
+}
+
+cache_available <- function(path) {
+  file.exists(path) || cache_allowed(path)
+}
+
+cache_allowed <- function(path) {
+  if (!interactive()) return(FALSE)
+
+  cat("Use a local file ('", path, "'), to cache OAuth access credentials ",
+      "between R sessions?\n", sep = "")
+  utils::menu(c("Yes", "No")) == 1
+}
+
+cache_create <- function(path) {
+
+  cache_parent <- dirname(path)
+  if (!dir.exists(cache_parent)) {
+    dir.create(cache_parent, recursive = TRUE)
+  }
+
+  file.create(path, showWarnings = FALSE)
+  if (!file.exists(path)) {
+    stop("Failed to create local cache ('", path, "')", call. = FALSE)
+  }
+
+  "!DEBUG cache exists: `path`"
+
+  ## owner can read and write, but not execute; no one else can do anything
+  Sys.chmod(path, "0600")
+
+  desc <- file.path(cache_parent, "DESCRIPTION")
+  if (file.exists(desc)) {
+    add_line(
+      file.path(cache_parent, ".Rbuildignore"),
+      paste0("^", gsub("\\.", "\\\\.", path), "$")
+    )
+  }
+  git <- file.path(cache_parent, c(".gitignore", ".git"))
+  if (any(file.exists(git))) {
+    add_line(
+      file.path(cache_parent, ".gitignore"),
+      path
+    )
+  }
+
+  TRUE
+}
+
+cache_read <- function(path) {
+  if (file.size(path) == 0) {
+    list()
+  } else {
+    validate_token_list(readRDS(path))
+  }
+}
+
+cache_write <- function(tokens, path) {
+  saveRDS(tokens, path)
+}
+
+validate_token_list <- function(existing) {
+  hashes <- vapply(existing, function(x) x$hash(), character(1), USE.NAMES = FALSE)
+  nms <- names(existing)
+
+  if (!identical(nms, hashes)) {
+    stop("Token names do not match their hash!", call. = FALSE)
+  }
+
+  if (anyDuplicated(nms)) {
+    stop("Duplicate tokens!, call. = FALSE")
+  }
+
+  existing
+}
+
+## useful to jennybc during development
+cache_show <- function(path = NULL) { # nocov start
```

Comment:
Maybe eventually make `cache_sitrep()` and export?

Reply:
Yeah, something like that is where this heading. You will also need to be able to clear your cache or make targeted deletions.